### PR TITLE
SqlRecipientList Management

### LINKED ISSUE #105: Recipient SQL Management UI
## Is your feature request related to a problem? Please describe

Current design allows for a recipient list to have a sql query defined, but this doesn't provide a nice way to parameterise, validate and use queries on the backend in notifications.

## Describe the solution you'd like

A UI for managing SQL Queries that create recipeints
Backend validations of queries

### PR DESCRIPTION
**Pull request by jmbrunskill**
Closes #105 

# 👩🏻‍💻 What does this PR do?
 

Users are able to create queries for new Recipients from the configured datasource.
They can test their queries to see if they have the right format

## Testing a query with no parameters:

Kapture 2023-09-07 at 18 43 37

Parameters can be added in tera format e.g. `{{ parameter }}`

## Testing a query with parameters:
Kapture 2023-09-07 at 18 45 09

## New parameters are automatically detected...

Kapture 2023-09-07 at 18 47 15

## If you don't return the right columns you get an error...

Kapture 2023-09-07 at 18 49 56

## If you return an invalid `notification_type` you get an `Unknown` 
Kapture 2023-09-07 at 18 52 05

> It's hard to return the actual value when it's diesel enum, could be mapped in service layer probably which could be a nicer solution?

## Queries are set to readonly:

Kapture 2023-09-07 at 18 48 16

# 🧪 How has/should this change been tested?

## 💌 Any notes for the reviewer?

As this is already a massive PR, I'm thinking it is probably best if we make new issues for UX/UI Improvements.

### CODE REVIEW

File: backend/datasource/src/recipient.rs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    pub notification_type: String,
+    #[diesel(sql_type = Text)]
+    pub to_address: String,
+}
+
+pub fn pg_sql_query_as_recipients(
+    connection: &mut PgConnection,
+    sql_select_query: String,
+) -> Result<Vec<BasicRecipientRow>, DieselError> {
+    let _ = connection.batch_execute("SET SESSION CHARACTERISTICS AS TRANSACTION READ ONLY;")?;
```
Comment (clemens-msupply) [inline backend/datasource/src/recipient.rs:25]: Could you please explain why do you do this? e.g. vs a normal transaction?

**Review by clemens-msupply — APPROVED** [review]
Looks good! mostly skipped over FE changes... Just some inline nits...

File: backend/datasource/src/recipient.rs (hunk 1 above)
Comment (jmbrunskill) [reply backend/datasource/src/recipient.rs:25]: As discussed, this is to provide a read only transaction. It reduced the risk of destructive activities or changing data via an sql query.

File: backend/graphql/recipient_list/src/mutations/create_sql.rs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    }
+}
+
+#[derive(InputObject, Clone)]
+pub struct CreateSqlRecipientListInput {
+    pub id: String,
+    pub name: String,
+    pub description: String,
+    pub query: String,
+    pub parameters: String,
```
[the remark this answers is not in the document]
Comment (jmbrunskill) [reply backend/graphql/recipient_list/src/mutations/create_sql.rs:43]: Agreed updated...

File: backend/graphql/recipient_list/src/mutations/create_sql.rs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        )),
+        Err(error) => map_sql_error(error),
+    }
+}
+
+#[derive(InputObject, Clone)]
+pub struct CreateSqlRecipientListInput {
+    pub id: String,
+    pub name: String,
+    pub description: String,
```
[the remark this answers is not in the document]
Comment (jmbrunskill) [reply backend/graphql/recipient_list/src/mutations/create_sql.rs:41]: I'm happy with it being required. Can be set to empty string but avoid extra optionals throughout the code.
Possibly could remove it entirely, but I think it's worth it...

File: backend/service/src/sql_recipient_list/create.rs [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    StorageConnection,
+};
+
+#[derive(Clone, Default)]
+pub struct CreateSqlRecipientList {
+    pub id: String,
+    pub name: String,
+    pub description: String,
+    pub query: String,
+    pub parameters: String,
```
[the remark this answers is not in the document]
Comment (jmbrunskill) [reply backend/service/src/sql_recipient_list/create.rs:24]: Great idea. I'll update this.

**Pull request merged**